web/src/message_fetch.js

```diff
@@ -61,11 +61,23 @@ function process_result(data, opts) {
         }
     }
 
-    if (messages.length > 0 && (opts.msg_list === message_lists.home || opts.is_recent_view_data)) {
+    if (
+        messages.length > 0 &&
+        // To avoid non-contiguous blocks of data in recent view
+        // from message_lists.home and recent_view_message_list_data,
+        // we only process data from message_lists.home if we have
+        // found the newest message in message_lists.home.
+        ((opts.msg_list === message_lists.home && has_found_newest) || opts.is_recent_view_data)
+    ) {
         // We keep track of how far back we've fetched messages in
         // the recent view. This assumes `messages` is already sorted.
-        const oldest_timestamp = all_messages_data.first().timestamp;
         recent_view_ui.process_messages(messages);
+        let oldest_timestamp;
+        if (opts.is_recent_view_data) {
+            oldest_timestamp = opts.msg_list_data.first().timestamp;
+        } else {
+            oldest_timestamp = all_messages_data.first().timestamp;
+        }
```

Comment:
`oldest_timestamp` could be incorrect here for like 1 fetch when there are less than 200 unmuted unreads. But, it will be correct in next fetch. Not sure it is worth writing extra logic for it.

Reply:
I think you still need a call to `recent_view_ui.process_messages();` call passing all the messages that have been accumulated in `all_messages_data` to occur somewhere the first time that `has_found_newest` is true -- that call can probably come from `home_view_loaded` or something rather than here, though. 

BTW I wonder if after the previous commit, `recent_view_ui.set_oldest_message_date` should be inlined into `recent_view_ui.process_messages` or something.

Reply:
Yeah, both make sense to do. Updated the PR.